src/extra/components/dld.py

```diff
@@ -0,0 +1,389 @@
+                finite_mask.reshape(-1, num_rows).sum(axis=1)).to_frame()
+
+            # Insert additional index level enumerating actual entries.
+            if entry_level is not None:
+                index_df[entry_level] = np.flatnonzero(finite_mask) % num_rows
+
+            res.append(pd_cls(
+                raw[finite_mask], pd.MultiIndex.from_frame(index_df)))
+
+        return res if split_axis is not None else res[0]
+
+    @property
+    def detector_name(self):
+        return self._detector_name
+
+    @property
+    def control_source(self):
+        """Control source."""
+
+        if self._control_src is None:
+            raise ValueError('component is initialized with earlier data '
+                             'lacking a control source')
+
+        return self._control_src
+
+    @property
+    def instrument_source(self):
+        """Instrument source."""
+        return self._instrument_src
+
+    @property
+    def rec_params(self):
+        """Reconstruction parameters."""
+        return self.control_source.run_values()
+
+    def select_trains(self, trains):
+        new_self = copy(self)
+
+        if self._control_src is not None:
+            new_self._control_src = self._control_src.select_trains(trains)
+        new_self._instrument_src = self._instrument_src.select_trains(trains)
+        new_self._pulses = self._pulses.select_trains(trains)
+
+        return new_self
+
+    def pulses(self):
+        """Get pulse object based on internal triggers.
+
+        Returns:
+            (extra.components.DldPulses): Pulse object based on
+                constructed trigger information.
+        """
+
+        from .pulses import DldPulses
+
+        if isinstance(self._pulses, DldPulses):
+            return self._pulses
+
+        return DldPulses(self._instrument_src)
+
+    def triggers(self):
+        """Get triggers as dataframe.
+
+        Returns:
+            (pandas.DataFrame): Constructed trigger information as
+                dataframe.
+        """
+
+        return self.pulses().triggers()
+
+    def edges(self, channel_index=True, pulse_dim='pulseId'):
+        """Get raw edges as series or dataframe.
+
+        Args:
+            channel_index (bool, optional): Whether to insert the edge
+                channel as index level and return Series object
+                (default), or as column and return DataFrame object.
+            pulse_dim ({pulseId, pulseIndex, time}, optional): Label
+                for pulse dimension, pulse ID by default.
+
+        Returns:
+            (pd.Series or pd.DataFrame): Raw edge positions as Series
+                (indexed by channel) or DataFrame (with channel as
+                column) object.
+        """
+
+        kd = self._instrument_src['raw.edges']
+        index = self._align_pulse_index(kd, pulse_dim)
+        data = kd.ndarray()
+        raw_edges = [self._build_reduced_pd(data[:, i, :], index, None)
+                     for i in range(7)]
```

Comment:
Without knowing anything about it, the hardcoded 7 feels a bit odd.  Are we confident it can never change? Does `data.shape[1]` work?

Reply:
Essentially for the moment yes, as it's hardcoded in the REMI/DLD notebook. But your suggestion makes sense, thanks!